networkmanager/networkmanager_test.go

```diff
@@ -71,542 +114,944 @@ func TestMain(m *testing.M) {
+	networkmanager.GetIPSubnet = getIPSubnet
+
+	defer func() {
+		networkmanager.GetIPSubnet = nil
+	}()
+
+	manager, err := networkmanager.New(&config.Config{}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer manager.Close()
+
+	for _, item := range testData {
+		if err := manager.AddInstanceToNetwork("instance0", "network0", item.params); err != nil {
+			t.Fatalf("Can't add instance to network: %s", err)
+		}
 
-	numNetworks := 3
-	numServices := 3
+		if string(cniInterface.networkConfig.Bytes) != item.networkConfig {
+			t.Errorf("Wrong network config: %s", string(cniInterface.networkConfig.Bytes))
+		}
 
-	for j := 0; j < numNetworks; j++ {
-		for i := 0; i < numServices; i++ {
-			if err := manager.AddServiceToNetwork(fmt.Sprintf("service%d%d", j, i),
-				fmt.Sprintf("network%d", j), networkmanager.NetworkParams{}); err != nil {
-				t.Fatalf("Can't add service to network: %s", err)
-			}
+		if err := manager.RemoveInstanceFromNetwork("instance0", "network0"); err != nil {
+			t.Fatalf("Can't remove instance from network: %s", err)
 		}
 	}
+}
 
-	if err := manager.DeleteAllNetworks(); err != nil {
-		t.Fatalf("Can't delete all networks: %s", err)
+func TestBandwithPlugin(t *testing.T) {
+	testData := []testPluginsData{
+		{
+			params: networkmanager.NetworkParams{
+				IngressKbit: 1200,
+				EgressKbit:  1200,
+			},
+			networkConfig: createPlugins([]string{
+				createBridgePlugin(""),
+				createBandwithPlugin(1200000, 1200000),
+				createDNSPlugin(),
+			}),
+		},
+		{
+			params: networkmanager.NetworkParams{
+				IngressKbit: 400,
+				EgressKbit:  300,
+			},
+			networkConfig: createPlugins([]string{
+				createBridgePlugin(""),
+				createBandwithPlugin(400000, 300000),
+				createDNSPlugin(),
+			}),
+		},
+	}
+
+	cniInterface := &testCNIInterface{}
+
+	networkmanager.CNIPlugins = cniInterface
+	networkmanager.GetIPSubnet = getIPSubnet
+
+	defer func() {
+		networkmanager.GetIPSubnet = nil
+	}()
+
+	manager, err := networkmanager.New(&config.Config{}, nil)
+	if err != nil {
+		t.Fatal(err)
 	}
+	defer manager.Close()
+
+	for _, item := range testData {
+		if err := manager.AddInstanceToNetwork("instance0", "network0", item.params); err != nil {
+			t.Fatalf("Can't add instance to network: %s", err)
+		}
 
-	for j := 0; j < numNetworks; j++ {
-		for i := 0; i < numServices; i++ {
-			if err := manager.IsServiceInNetwork(fmt.Sprintf("service%d%d", j, i),
-				fmt.Sprintf("network%d", j)); err == nil {
-				t.Fatalf("Service should not be in network: %s", err)
-			}
+		if string(cniInterface.networkConfig.Bytes) != item.networkConfig {
+			t.Errorf("Wrong network config: %s", string(cniInterface.networkConfig.Bytes))
+		}
+
+		if err := manager.RemoveInstanceFromNetwork("instance0", "network0"); err != nil {
+			t.Fatalf("Can't remove instance from network: %s", err)
 		}
 	}
 }
 
-func TestInternet(t *testing.T) {
-	t.Cleanup(func() { _ = manager.DeleteNetwork("network0") })
+func TestTrafficMonitoring(t *testing.T) {
```

Comment:
Consider to use table in this test. You have many similar actions.

Reply:
Done